Added show extra_specs for flavors, code for default flavor to load from...

schivuk opened the pull request:
... config file, code to boot instance using port speed and unit tests for all and servers

Change-Id: I951c6be5ed8604a958b4f4413ceac6415f1520fd

Commit:
Added load flavors from config file + extra specs

Change-Id: I951c6be5ed8604a958b4f4413ceac6415f1520fd

schivuk commented:
Anyone know how to change the length of the commit-title?

sudorandom commented:
@schivuk http://stackoverflow.com/questions/927358/undo-the-last-git-commit or http://stackoverflow.com/questions/179123/edit-an-incorrect-commit-message-in-git

schivuk commented:
- Added support to nova flavor-list to show extra_specs for flavors and also let user boot an instance using a flavor with port speed specified. 
- Removed hardcoded flavor list from code and added support to load default flavors from config file
- Unit tests

sudorandom commented on jumpgate/compute/drivers/sl/flavors.py:
```diff
@@ -133,8 +62,77 @@ def on_get(self, req, resp, tenant_id=None):
+    _flavors = None
+
+    @classmethod
+    def get_flavors(cls):
```
This loading should be done inside of `jumpgate/compute/drivers/sl/__init__.py` and the flavor details (or a class wrapping access to the flavors) should be passed in explicitly to each route that needs it. That way, all loading is done at the beginning of the application and dependencies are more explicit, which leads to easier testing. An example of this exists for the endpoint listing for the sl identity driver.

sudorandom commented on jumpgate/compute/drivers/sl/servers.py:
```diff
@@ -25,6 +30,7 @@ def __init__(self, app):
         self.app = app
 
     def on_post(self, req, resp, tenant_id, instance_id):
+        FLAVORS = flavors.Flavors.get_flavors()
```
Since this isn't a constant, FLAVORS should be lower cased.

sudorandom commented on etc/jumpgate.conf:
```diff
@@ -54,3 +54,6 @@ driver=jumpgate.network.drivers.sl
 driver=jumpgate.baremetal.drivers.sl
+
+[flavors]
+flavor_list='{"1": {"disk-type": "SAN", "name": "1 vCPU, 1GB ram, 25GB, SAN", "ram": 1024, "cpus": 1, "disk": 25, "id": "1", "portspeed": 100}, "2": {"disk-type": "SAN", "name": "1 vCPU, 1GB ram, 100GB, SAN", "ram": 1024, "cpus": 1, "disk": 100, "id": "2", "portspeed": 100}, "3": {"disk-type": "SAN", "name": "2 vCPU, 2GB ram, 100GB, SAN", "ram": 2048, "cpus": 2, "disk": 100, "id": "3", "portspeed ...
```
This could be a path to a json document with this same information. That way, you can actually format the JSON to be somewhat readable. There's an example of this (although not with JSON) in the sl identity driver.

schivuk replied:
Just want to confirm, is it like placing the JSON string in an external file like identity.templates in jumpgate.conf? And will it contain the entire JSON string with proper format and then my init file would load the file and convert its contents into a JSON dict?

sudorandom replied:
Yeah, I think we're on the same page.

For the service listing, the token driver is actually loaded in each **init** function (https://github.com/softlayer/jumpgate/blob/master/jumpgate/identity/drivers/sl/tokens.py#L144), but it probably makes much more sense to do it in the `jumpgate/compute/drivers.sl/__init__.py` file. For identity, the line I linked above would appear here: https://github.com/softlayer/jumpgate/blob/master/jumpgate/identity/drivers/sl/__init__.py#L39 and the fully loaded object would be passed in (instead of the filename like exists right now)

Commit:
Merge branch 'master' into totalW

Conflicts:
	jumpgate/compute/drivers/sl/servers.py
	tests/jumpgate-tests/compute/test_servers.py

schivuk commented:
- Moved the loading of flavors into the init file and passed the flavors into each route
- Added a json document to hold the flavors, formatted in a readable way
- Fixed other small issues as suggested

sudorandom commented on jumpgate/compute/drivers/sl/flavor_list_loader.py:
```diff
@@ -0,0 +1,99 @@
+
+                with open(json_file) as jf:
+                    json_str = jf.read()
+                    json_str = json_str.replace('\n', '')
```
Removing new lines and carridge returns is un-needed... I've demonstrated it in another PR, but here's an example again (now with more \r and \n):

``` python
import json

s = """{\n\r\n\r\r\n
    "firstName": "John",
    "lastName": "Smith",\n\r\n\r\r\n
    "isAlive": true,\n\r\n\r\r\n
    "age": 25,\n\r\n\r\r\n
    "height_cm": 167.6,\n\r\n\r\r\n
    "address":\n\r\n\r\r\n {
        "streetAddress": "21 2nd Street",
        \n\r\n\r\r\n"city": "New York",
        "state": "NY",
        "postalCode":\n\r\n\r\r\n "10021-3100"
    },
    "phoneNumbers": [\n\r\n\r\r\n
        { "type": "home", "number": "212 555-1234" },\n\r\n\r\r\n
        { "type": "office",  "number": "646 555-4567" }\n\r\n\r\r\n
    ]\n\r\n\r\r\n
}"""

print json.loads(s)
```

Commit:
Merge branch 'master' into totalW

Conflicts:
	jumpgate/compute/drivers/sl/servers.py

schivuk commented:
- Removed the unnecessary 'replace' lines

jimlindeman commented on jumpgate/compute/drivers/sl/extra_specs.py:
```diff
@@ -1,5 +1,51 @@
+        '''Returns the extra specs for a particular flavor
+
+        '''
+        all_flavors = flavor_list_loader.Flavors.get_flavors(mock.MagicMock())
```
I may be confused here, but why is mainline code referencing mock.MackMock() instead of only in the test code paths?

schivuk replied:
Sorry, was a mistake on my part. I've made the fix and passed the flavor list through those 2 routes in init file.

Commit:
Resolved the MagicMock issue

Commit:
Merge branch 'master' into totalW

Conflicts:
	tests/jumpgate-tests/common/test_dispatchers.py
	tests/jumpgate-tests/test.jumpgate.conf

jimlindeman commented:
+1
Looks better now with the fixed Mock references passing the loaded flavors now.
